File: packages/sdk-csharp/src/Theatrical.Sdk/Types/Film.cs
namespace Theatrical.Sdk.Types;

public sealed record Film
{
    public required string Id { get; init; }
    public required string Title { get; init; }
    public string? Synopsis { get; init; }
    public string[]? Genres { get; init; }
    public string? Rating { get; init; }
    public int? RuntimeMinutes { get; init; }
    public int? Runtime { get; init; }
    public string? PosterUrl { get; init; }
    public string? ReleaseDate { get; init; }
    public string? Director { get; init; }
    public bool IsNowShowing { get; init; }
    public bool IsComingSoon { get; init; }
    public CastMember[]? Cast { get; init; }
}

public sealed record FilmDetail
{
    public required string Id { get; init; }
    public required string Title { get; init; }
    public string? Synopsis { get; init; }
    public string[]? Genres { get; init; }
    public string? Rating { get; init; }
    public int? RuntimeMinutes { get; init; }
    public string? PosterUrl { get; init; }
    public string? ReleaseDate { get; init; }
    public string? Director { get; init; }
    public CastMember[]? Cast { get; init; }
    public CrewMember[]? Crew { get; init; }
    public FilmRating[]? Ratings { get; init; }
    public string[]? Formats { get; init; }
    public string[]? Languages { get; init; }
}

public sealed record CastMember
{
    public required string Name { get; init; }
    public string? Role { get; init; }
}

public sealed record CrewMember
{
    public required string Name { get; init; }
    public required string Department { get; init; }
    public string? Job { get; init; }
}

public sealed record FilmRating
{
    public required string Source { get; init; }
    public required string Value { get; init; }
}

public sealed record FilmFilter
{
    public string? SiteId { get; init; }
    public string? Genre { get; init; }
    public string? Query { get; init; }
    public bool? NowShowing { get; init; }
    public bool? ComingSoon { get; init; }
}

public sealed record FilmSearchFilter
{
    public string? SiteId { get; init; }
    public string? Genre { get; init; }
    public string? Query { get; init; }
    public bool? NowShowing { get; init; }
    public bool? ComingSoon { get; init; }
    public int? Limit { get; init; }
    public int? Offset { get; init; }
    public string? RatingClassification { get; init; }
    public string? Format { get; init; }
    public string? Language { get; init; }
    public string? ReleaseDateFrom { get; init; }
    public string? ReleaseDateTo { get; init; }
    public int? MinRuntime { get; init; }
    public int? MaxRuntime { get; init; }
    public string? SortBy { get; init; }
    public string? SortOrder { get; init; }
}

File: packages/sdk-py/theatrical/types/film.py
"""Film types."""

from __future__ import annotations

from enum import Enum
from typing import Literal, Optional

from theatrical.types.base import ApiModel


class Genre(str, Enum):
    ACTION = "action"
    ADVENTURE = "adventure"
    ANIMATION = "animation"
    COMEDY = "comedy"
    CRIME = "crime"
    DOCUMENTARY = "documentary"
    DRAMA = "drama"
    FAMILY = "family"
    FANTASY = "fantasy"
    HORROR = "horror"
    MUSICAL = "musical"
    MYSTERY = "mystery"
    ROMANCE = "romance"
    SCI_FI = "sci-fi"
    THRILLER = "thriller"
    WAR = "war"
    WESTERN = "western"


class FilmFormat(str, Enum):
    TWO_D = "2D"
    THREE_D = "3D"
    IMAX = "IMAX"
    IMAX_3D = "IMAX 3D"
    FOUR_DX = "4DX"
    DOLBY_ATMOS = "Dolby Atmos"
    SCREENX = "ScreenX"


class FilmLanguage(str, Enum):
    EN = "en"
    ES = "es"
    FR = "fr"
    DE = "de"
    JA = "ja"
    KO = "ko"
    ZH = "zh"
    HI = "hi"
    TE = "te"
    MI = "mi"


class Rating(ApiModel):
    classification: str
    description: Optional[str] = None


class CastMember(ApiModel):
    name: str
    role: Optional[str] = None


class CrewMember(ApiModel):
    name: str
    department: str
    job: str


class FilmRating(ApiModel):
    source: str
    score: str
    out_of: Optional[str] = None


class Film(ApiModel):
    id: str
    title: str
    synopsis: str
    genres: list[Genre]
    runtime: int
    rating: Rating
    release_date: str
    poster_url: Optional[str] = None
    trailer_url: Optional[str] = None
    cast: list[CastMember]
    director: str
    distributor: Optional[str] = None
    is_now_showing: bool
    is_coming_soon: bool


class FilmDetail(Film):
    crew: list[CrewMember]
    ratings: list[FilmRating]
    formats: list[FilmFormat]
    languages: list[FilmLanguage]
    original_title: Optional[str] = None
    production_countries: Optional[list[str]] = None
    budget: Optional[float] = None
    box_office: Optional[float] = None
    website: Optional[str] = None


class FilmFilter(ApiModel):
    site_id: Optional[str] = None
    genre: Optional[Genre] = None
    query: Optional[str] = None
    now_showing: Optional[bool] = None
    coming_soon: Optional[bool] = None
    limit: Optional[int] = None
    offset: Optional[int] = None


class FilmSearchFilter(FilmFilter):
    rating_classification: Optional[str] = None
    format: Optional[FilmFormat] = None
    language: Optional[FilmLanguage] = None
    release_date_from: Optional[str] = None
    release_date_to: Optional[str] = None
    min_runtime: Optional[int] = None
    max_runtime: Optional[int] = None
    sort_by: Optional[Literal["title", "releaseDate", "runtime", "popularity"]] = None
    sort_order: Optional[Literal["asc", "desc"]] = None
